unison-cli/src/Unison/Cli/Monad.hs

```diff
@@ -0,0 +1,349 @@
+{-# LANGUAGE DataKinds #-}
+
+-- | The main CLI monad.
+module Unison.Cli.Monad
+  ( -- * Cli monad
+    Cli,
+    ReturnType (..),
+    runCli,
+
+    -- * Envronment
+    Env (..),
+
+    -- * Mutable state
+    LoopState (..),
+    loopState0,
+    getLoopState,
+    putLoopState,
+    modifyLoopState,
+
+    -- * Lifting IO actions
+    ioE,
+
+    -- * Acquiring resources
+    with,
+    withE,
+
+    -- * Scoping actions
+    newBlock,
+
+    -- * Short-circuiting
+    returnWith,
+    returnEarly,
+    returnEarlyWithoutOutput,
+    haltRepl,
+
+    -- * Communicating output to the user
+    respond,
+    respondNumbered,
+
+    -- * Debug-timing actions
+    time,
+
+    -- * Misc types
+    LoadSourceResult (..),
+  )
+where
+
+import Control.Lens (lens, set)
+import Control.Monad.Reader (MonadReader (..), ReaderT (ReaderT))
+import Control.Monad.Trans.Cont
+import qualified Data.Configurator.Types as Configurator
+import Data.Generics.Labels ()
+import Data.IORef
+import qualified Data.List.NonEmpty as List (NonEmpty)
+import qualified Data.List.NonEmpty as List.NonEmpty
+import Data.Time.Clock (DiffTime, diffTimeToPicoseconds)
+import Data.Time.Clock.System (getSystemTime, systemToTAITime)
+import Data.Time.Clock.TAI (diffAbsoluteTime)
+import GHC.OverloadedLabels (IsLabel (..))
+import System.CPUTime (getCPUTime)
+import Text.Printf (printf)
+import Unison.Auth.CredentialManager (CredentialManager)
+import Unison.Auth.HTTPClient (AuthenticatedHttpClient)
+import Unison.Codebase (Codebase)
+import Unison.Codebase.Branch (Branch)
+import Unison.Codebase.Editor.Input (Input)
+import Unison.Codebase.Editor.Output (NumberedArgs, NumberedOutput, Output)
+import Unison.Codebase.Editor.UCMVersion (UCMVersion)
+import qualified Unison.Codebase.Path as Path
+import Unison.Codebase.Runtime (Runtime)
+import qualified Unison.Debug as Debug
+import qualified Unison.Parser as Parser
+import Unison.Parser.Ann (Ann)
+import Unison.Prelude
+import qualified Unison.Server.CodebaseServer as Server
+import Unison.Symbol (Symbol)
+import qualified Unison.UnisonFile as UF
+
+-- | The main command-line app monad.
+--
+-- * It is a reader monad: get the 'Env' environment with 'ask'.
+--
+-- * It is a state monad (kind of): get the 'LoopState' with 'getLoopState', and put it with 'putLoopState'. The state
+-- is stored in an @IORef@, so state changes are visible across threads, and persist even when an exception is raised.
```

Comment:
This makes me a little nervous, I think it just means we'll have to be very careful about getting the root codebase in threads and such since it could change at any point without notice.

If an IORef allows us to use MonadUnliftIO though that might be nice?